### PR TITLE
feature: Test updateDelegations

### PR DESCRIPTION
**Pull request by gbarkhatov**
- `updateDelegations` is a pure function
- tests for :
  - different
  - same
  - empty
  - identical

Closes #282

### CODE REVIEW

File: src/utils/local_storage/updateDelegations.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { Delegation } from "@/app/types/delegations";
+
+import { filterDelegationsLocalStorage } from "./filterDelegationsLocalStorage";
+
+export const updateDelegations = async (
```
Comment (vitsalis) [inline src/utils/local_storage/updateDelegations.ts:5]:
```suggestion
export const calculateDelegationsDiff = async (
```

This does not do any updating.

File: src/utils/local_storage/updateDelegations.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  delegations: Delegation[],
+  delegationsLocalStorage: Delegation[],
+): Promise<{ areDelegationsDifferent: boolean; delegations: Delegation[] }> => {
+  // Filter the delegations that are still valid
+  const validDelegations = await filterDelegationsLocalStorage(
```
Comment (vitsalis) [inline src/utils/local_storage/updateDelegations.ts:10]:
```suggestion
  const validLocalStorageDelegations = await filterDelegationsLocalStorage(
```

File: src/utils/local_storage/updateDelegations.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      (hash, index) => hash !== delegationsLocalStorageHashes[index],
+    );
+
+  return {
+    areDelegationsDifferent,
+    // Return the new delegations if they are different or original if no update is needed
+    delegations: areDelegationsDifferent
+      ? validDelegations
+      : delegationsLocalStorage,
+  };
```
Comment (vitsalis) [inline src/utils/local_storage/updateDelegations.ts:36]: Don't understand why we are doing the above diff between `validDelegations` and `delegationsLocalStorage`. If they are different, we return `validDelegations`. If they are the same, we return `delegationsLocalStorage`. Why not return `validDelegations` at all times? Won't we get the same result?

If we are doing this just to know whether we should call `setDelegationsLocalStorage`, then (1) is the overhead of `setDelegationsLocalStorage` worth the additional complexity and (2) this seems a bit of spaghetti code for a utility to do something random completely unrelated to its purpose for a function up the stack. If we want to do this, we should have two utilities: (1) `filterLocalStorageDelegations(delegationsLocalStorage, delegationsAPI)` and (2) `delegationsDifferent(delegations1, delegations2): bool`.

File: tests/utils/local_storage/updateDelegations.test.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+// Mock the filterDelegationsLocalStorage function
+jest.mock("@/utils/local_storage/filterDelegationsLocalStorage");
+
+describe("updateDelegations", () => {
+  const mockDelegationsHash1: Delegation[] = [
+    { stakingTxHashHex: "hash1" } as Delegation,
+  ];
+  const mockDelegationsHash2: Delegation[] = [
+    { stakingTxHashHex: "hash2" } as Delegation,
+  ];
```
Comment (vitsalis) [inline tests/utils/local_storage/updateDelegations.test.ts:14]: We should have more elements in each list. Comparing one by one does not cover a lot of test cases. I recommend we have multiple lists, e.g. some with 1 element, others with many etc.

File: tests/utils/local_storage/updateDelegations.test.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  it("should handle both empty delegations and local storage correctly", async () => {
+    (filterDelegationsLocalStorage as jest.Mock).mockResolvedValue([]);
+
+    const result = await updateDelegations([], []);
+
+    expect(result.areDelegationsDifferent).toBe(false);
+    expect(result.delegations).toEqual([]);
+  });
+
+  it("should handle identical delegations correctly", async () => {
```
Comment (vitsalis) [inline tests/utils/local_storage/updateDelegations.test.ts:78]: This test has already been performed.

File: src/utils/local_storage/updateDelegations.ts (hunk 2 above)
Comment (gbarkhatov) [reply src/utils/local_storage/updateDelegations.ts:10]: Done

File: src/utils/local_storage/updateDelegations.ts (hunk 1 above)
Comment (gbarkhatov) [reply src/utils/local_storage/updateDelegations.ts:5]: Done

File: tests/utils/local_storage/updateDelegations.test.ts (hunk 5 above)
Comment (gbarkhatov) [reply tests/utils/local_storage/updateDelegations.test.ts:78]: Done

File: src/utils/local_storage/updateDelegations.ts (hunk 3 above)
Comment (gbarkhatov) [reply src/utils/local_storage/updateDelegations.ts]: Done

**Review by vitsalis — APPROVED** [review]

File: src/app/page.tsx [hunk 6]
```diff
@@ -287,36 +287,18 @@ const Home: React.FC<HomeProps> = () => {
...
-        .map((delegation) => delegation.stakingTxHashHex)
-        .sort();
-
-      // Check if the validDelegations are different from the current delegationsLocalStorage
-      const areDelegationsDifferent =
-        validDelegationsHashes.length !==
-          delegationsLocalStorageHashes.length ||
-        validDelegationsHashes.some(
-          (hash, index) => hash !== delegationsLocalStorageHashes[index],
+    const update = async () => {
```
[the remark this answers is not in the document]
Comment (gbarkhatov) [reply src/app/page.tsx]: Okay

**Pull request merged**